src/main/java/com/iota/iri/service/milestone/impl/MilestoneSolidifierImpl.java

```diff
@@ -1,339 +1,666 @@
+                if (firstRun) {
+                    firstRun = false;
+                    bootStrapSolidMilestones();
+                    ledgerService.restoreLedgerState();
+
+                    Snapshot latestSnapshot = snapshotProvider.getLatestSnapshot();
+                    setLatestMilestone(latestSnapshot.getHash(), latestSnapshot.getIndex());
+                    logChange(snapshotProvider.getInitialSnapshot().getIndex());
+
+                    syncProgressInfo.setSyncMilestoneStartIndex(snapshotProvider.getInitialSnapshot().getIndex());
+                }
+
+                solidifySeenMilestones();
+                processSolidifyQueue();
+                checkLatestSolidMilestone();
+
+                if (latestMilestoneIndex > latestSolidMilestone) { solidifyLog();}
+            } catch(Exception e) {
+                log.error("Error running milestone solidification thread", e);
+            }
+        }
+    }
 
     /**
-     * Holds all unsolid milestones that shall be solidified (the transaction hash mapped to its milestone index).
+     * Scan through milestones in the {@link #unsolidMilestones} queue and refill the {@link #solidificationQueue} queue.
+     * It then iterates through this queue to determine iff the milestone has already been seen and validated. If it is
+     * found in the {@link #seenMilestones} queue, then the milestone is removed from the solidification pool. If not,
+     * then the {@link #oldestMilestoneInQueue} is updated iterating through whatever milestones are still present in
+     * the {@link #solidificationQueue}.
+     *
+     * If upon scanning the {@link #unsolidMilestones} queue is empty and {@link #initialized} is false, set
+     * {@link #initialized} to true.
      */
-    private final Map<Hash, Integer> unsolidMilestonesPool = new ConcurrentHashMap<>();
+    private void scanMilestonesInQueue() {
+        // refill the solidification queue with solidification candidates sorted by index
+        if (unsolidMilestones.size() > 0) {
+            Map<Hash, Integer> newSolidificationQueue = new ConcurrentHashMap<>();
+            unsolidMilestones.entrySet().stream()
+                    .sorted(Map.Entry.comparingByValue())
+                    .forEach(milestone -> {
+                        if (newSolidificationQueue.size() < MAX_SIZE) {
+                            newSolidificationQueue.put(milestone.getKey(), milestone.getValue());
+                        }
+                    });
+            solidificationQueue = newSolidificationQueue;
+        }
+
+        // update the oldest milestone in queue
+        oldestMilestoneInQueue = null;
+        for (Map.Entry<Hash, Integer> currentEntry : solidificationQueue.entrySet()) {
+            if (seenMilestones.containsKey(currentEntry.getValue())) {
+                // remove candidate from pool
+                unsolidMilestones.remove(currentEntry.getKey());
+                solidificationQueue.remove(currentEntry.getKey());
```

Comment:
why are those values can be in the queue??
Maybe better to check upon insertion to the queues?

Reply:
I'm going to move it to the validation check, since that's the only place a transaction ends up in the seenMilestones pool. This was left over logic from when I had insertion to the seenMilestones pool from the `MilestoneStage`.